### PR TITLE
Make verlet lists a blackbox

### LINKED ISSUE #171: Verlet lists need to be black boxed
**Is your feature request related to a problem? Please describe.**
When using AutoPas over multiple MPI processes it is not possible to use different types of containers on different processes, e.g. Linked Cells on one process and Verlet Lists on the other. The current implementation of the verlet lists enforces this behavior as the particle interchange happens differently.

**Describe the solution you'd like**
Make AutoPas a proper black box, such that the particle interchanges work the same for each container. This can be done by either of the two options:
1. Build a verlet list for the inner cells that are not affected by the outside. And build a new list in every step for the outer cells.
2. Build a verlet list for the inner cells that are not affected by the outside. And iterate over the boundary parts separately.

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

Makes it possible to use VerletLists as a black box. 

- [x] Implement BlackBoxMode.
- [x] Make black box mode usable from outside. Integration in AutoPas interface
- [x] Check convenience functions (checkNeighborListsAreValid, isContainerUpdateNeeded, needsRebuild)
- [x] save type of verlet lists (newton3 or not) in VerletLists

## Resolved Issues

- [x] fixes #171: implements option 2.
- [x] fixes wrong verlet list results if newton3 is disabled.

# How Has This Been Tested?

- [x] Test for AoS - N3 and noN3
- [x] Test for SoA - N3 and noN3
- [x] Test with halo deletion
- [x] Tested within ls1

# Performance data:
for single thread's verlet lists remain faster than linked cells, even in blackbox mode:
(tested in ls1 mardyn)

| type              | verlet | verletBB | linkedCells |
|-------------------|--------|----------|-------------|
| single thread aos | 0.26   | 0.39     | 0.44        |
| single thread soa | 0.1    | 0.31     | 0.40        |

### CODE REVIEW

File: src/autopas/containers/cellPairTraversals/C08BasedTraversal.h [hunk 1]
```diff
@@ -47,25 +46,51 @@ class C08BasedTraversal : public CellPairTraversal<ParticleCell> {
...
+    LoopBody&& loopBody) {
+  if (blackBoxTraversalOption == outer) {
+    // tooSmall indicates whether there is any inner region.
+    // If there is no inner region, just do a normal traversal, as the traversals are supposedly the same.
+    bool tooSmall = false;
+    for (unsigned short d = 0; d < 3; ++d) {
+      tooSmall |= this->_cellsPerDimension[d] < 6;
+    }
+    if (not tooSmall) {
+      return c08TraversalOuter(loopBody);
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/C08BasedTraversal.h]: this is a VOID function!

File: src/autopas/containers/verletListsCellBased/verletLists/VerletListHelpers.h [hunk 2]
```diff
@@ -24,10 +24,12 @@ class VerletListHelpers {
   typedef std::unordered_map<Particle *, std::vector<Particle *>> AoS_verletlist_storage_type;
 
   /// typedef for soa's of verlet list's linked cells (only id and position needs to be stored)
-  typedef utils::SoAType<size_t, double, double, double>::Type SoAArraysType;
+  // typedef utils::SoAType<size_t, double, double, double>::Type SoAArraysType;
```
Comment (FG-TUM) [inline src/autopas/containers/verletListsCellBased/verletLists/VerletListHelpers.h]: still needed?

File: src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h [hunk 3]
```diff
@@ -56,20 +58,22 @@ class VerletLists
    * The neighbor lists are build using a search radius of cutoff + skin.
    * The rebuildFrequency should be chosen, s.t. the particles do not move more
    * than a distance of skin/2 between two rebuilds of the lists.
-   * @param boxMin the lower corner of the domain
-   * @param boxMax the upper corner of the domain
-   * @param cutoff the cutoff radius of the interaction
-   * @param skin the skin radius
-   * @param rebuildFrequency specifies after how many pair-wise traversals the
+   * @param boxMin The lower corner of the domain
```
Comment (FG-TUM) [inline src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h]:
```suggestion
   * @param boxMin The lower corner of the domain.
```
Also applies to the following lines :P

File: src/autopas/utils/StaticSelectorMacros.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+#pragma once
+
+/**
+ * Will execute the passed body (=variadic argument) with the static value of the bool.
+ * @param theBool The bool to be used.
+ * @note The second Argument is variadic such that commas pose no problem.
+ */
+#define AUTOPAS_WITH_STATIC_BOOL(theBool, ...) \
```
Comment (FG-TUM) [inline src/autopas/utils/StaticSelectorMacros.h]: convert to template function

File: src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h [hunk 5]
```diff
@@ -297,71 +372,267 @@ class VerletLists
...
+   * Loops over cells that are relevant for the inner traversal.
+   * @tparam LoopBody
+   * @param dims
+   * @param parallel define if this is allowed to be parallel
+   * @param loopBody
    */
-  size_t updateIdMapAoS() {
-    size_t i = 0;
+  template <class LoopBody>
+  void innerRelevantTraversal(const std::array<size_t, 3>& dims, bool parallel, LoopBody loopBody) {
```
Comment (FG-TUM) [inline src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h]:
```suggestion
  void innerTraversal(const std::array<size_t, 3>& dims, bool parallel, LoopBody loopBody) {
```
or is there also an irrelevantTraversal?

Comment (FG-TUM) [reply src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h]: also applies to similar function names

File: src/autopas/pairwiseFunctors/FlopCounterFunctor.h [hunk 6]
```diff
@@ -144,7 +144,7 @@ class FlopCounterFunctor : public Functor<Particle, ParticleCell> {
     double *const __restrict__ yptr = soa.template begin<Particle::AttributeNames::posY>();
     double *const __restrict__ zptr = soa.template begin<Particle::AttributeNames::posZ>();
 
-    for (unsigned int i = iFrom; i < iTo; ++i) {
+    for (unsigned long i = iFrom; i < iTo; ++i) {
```
Comment (FG-TUM) [inline src/autopas/pairwiseFunctors/FlopCounterFunctor.h]:
```suggestion
    for (size_t i = iFrom; i < iTo; ++i) {
```
?

File: src/autopas/utils/SoAStorage.h [hunk 7]
```diff
@@ -19,27 +19,51 @@ template <class SoAArraysType>
...
 
  public:
   /**
    * Apply the specific function to all vectors.
    * This can e.g. be resize operations, ...
    * The functor func should not require input parameters. The returns of the functor are ignored.
-   * @tparam FunctorT the type of the functor
-   * @param func a functor, that should be applied on all vectors (e.g. lambda functions, should take `auto& list` as an
-   * argument)
+   * @tparam FunctorT The type of the functor
```
Comment (FG-TUM) [inline src/autopas/utils/SoAStorage.h]:
```suggestion
   * @tparam FunctorT The type of the functor.
```

**Comment by SteffenSeckler** [comment]
will be implemented differently, by instead, increasing the size of linked cells

**Pull request closed without merging**